### PR TITLE
magnum: bump dependencies + modernize

### PR DESCRIPTION
**Pull request by SpaceIm**
- relocatable shared libs on macOS: see https://github.com/conan-io/hooks/issues/376
- `CMakeDeps` support
- cache CMake configuration with `functools.lru_cache`
- use target `glfw` instead of unofficial `glfw::glfw` target
- reorder methods by order of execution
- bump `vulkan-loader`

---

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/magnum/all/test_package/conanfile.py [hunk 1]
```diff
@@ -4,18 +4,22 @@
...
 
 class TestPackageConan(ConanFile):
-    settings = "os", "compiler", "build_type", "arch"
-    generators = "cmake", "cmake_find_package"
+    settings = "os", "arch", "compiler", "build_type"
+    generators = "cmake", "cmake_find_package_multi"
+
+    def build_requirements(self):
+        if hasattr(self, "settings_build"):
+            self.build_requires(str(self.requires["magnum"]))
```
Comment (jgsogo) [inline recipes/magnum/all/test_package/conanfile.py:12]: Two comments:
 + Why is this needed?
 + Although this strategy works in CCI, it is essentially wrong. It will require `magnum` for the _build_ context, which is a different configuration from the one that has been built. In a cross-building scenario, the command `conan create` will fail because `magnum` is not built for that configuration.
 
 ```suggestion
```

**Review by jgsogo — CHANGES_REQUESTED** [review]

File: recipes/magnum/all/test_package/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/magnum/all/test_package/conanfile.py:12]: Indeed it's not required since CMakeLists of test package don't test executable imported targets when cross-building, which is probably not very neat: https://github.com/conan-io/conan-center-index/blob/master/recipes/magnum/all/test_package/CMakeLists.txt#L30-L83

File: recipes/magnum/all/test_package/conanfile.py [hunk 2]
```diff
@@ -4,18 +4,19 @@
 
 
 class TestPackageConan(ConanFile):
-    settings = "os", "compiler", "build_type", "arch"
-    generators = "cmake", "cmake_find_package"
+    settings = "os", "arch", "compiler", "build_type"
+    generators = "cmake", "cmake_find_package_multi"
+
```
Comment (SpaceIm) [inline recipes/magnum/all/test_package/conanfile.py:9]:
```suggestion
```

**Review by SSE4 — APPROVED** [review]

File: recipes/magnum/all/test_package/conanfile.py (hunk 1 above)
Comment (jgsogo) [reply recipes/magnum/all/test_package/conanfile.py]: Yes, it is the same strategy and it is wrong. 

When you execute `conan create --profile:host=windows --profile:build=linux` you create packages for `windows`, and the `test_package/conanfile.py` will receive the same profile inputs. If the `test_package/conanfile.py` uses the package itself as build-requires, then it will be requesting a package for `linux` which hasn't been built. In `test_package/conanfile.py` we can only use/test the packages that have been built (`windows` ones).

In CCI this happens to work because we are building Intel-macos before building M1-macos. The cross-building scenario is `conan create --profile:host=m1 --profile:build=intel`. It works because `intel` has already been built, so the package for `intel` is available and it is found when used as build-requires.

If, for some reason, we decide to trigger `intel` and `m1` in parallel, or we do it the other way around (in the future we will be using M1 machines to build all MacOS packages) these pull-requests will fail. 

Locally, users running intel-macos would need to build first Intel-macos binaries (no cross-building) and then cross-build binaries for M1.... which can be unexpected if you are not aware of these issues.

**Review by jgsogo — APPROVED** [review]
